xds/src/main/java/io/grpc/xds/FilterChainMatchingProtocolNegotiators.java

```diff
@@ -0,0 +1,372 @@
+import io.grpc.Attributes;
+import io.grpc.internal.ObjectPool;
+import io.grpc.netty.GrpcHttp2ConnectionHandler;
+import io.grpc.netty.InternalProtocolNegotiationEvent;
+import io.grpc.netty.InternalProtocolNegotiator;
+import io.grpc.netty.InternalProtocolNegotiator.ProtocolNegotiator;
+import io.grpc.netty.ProtocolNegotiationEvent;
+import io.grpc.xds.EnvoyServerProtoData.CidrRange;
+import io.grpc.xds.EnvoyServerProtoData.ConnectionSourceType;
+import io.grpc.xds.EnvoyServerProtoData.FilterChain;
+import io.grpc.xds.EnvoyServerProtoData.FilterChainMatch;
+import io.grpc.xds.FilterChainMatchingProtocolNegotiators.FilterChainMatchingHandler.FilterChainSelector;
+import io.grpc.xds.internal.Matchers.CidrMatcher;
+import io.grpc.xds.internal.sds.SslContextProviderSupplier;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.util.AsciiString;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+
+/**
+ * Handles L4 filter chain match for the connection based on the xds configuration.
+ * */
+final class FilterChainMatchingProtocolNegotiators {
+  private static final Logger log = Logger.getLogger(
+          FilterChainMatchingProtocolNegotiators.class.getName());
+
+  private static final AsciiString SCHEME = AsciiString.of("http");
+
+  private FilterChainMatchingProtocolNegotiators() {
+  }
+
+  @VisibleForTesting
+  static final class FilterChainMatchingHandler extends ChannelInboundHandlerAdapter {
+
+    private final GrpcHttp2ConnectionHandler grpcHandler;
+    private final FilterChainSelector selector;
+    private final ProtocolNegotiator delegate;
+
+    FilterChainMatchingHandler(
+            GrpcHttp2ConnectionHandler grpcHandler, FilterChainSelector selector,
+            ProtocolNegotiator delegate) {
+      this.grpcHandler = checkNotNull(grpcHandler, "grpcHandler");
+      this.selector = checkNotNull(selector, "selector");
+      this.delegate = checkNotNull(delegate, "delegate");
+    }
+
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+      if (!(evt instanceof ProtocolNegotiationEvent)) {
+        super.userEventTriggered(ctx, evt);
+        return;
+      }
+      SelectedConfig config;
+      try {
+        config = selector.select(
+                (InetSocketAddress) ctx.channel().localAddress(),
+                (InetSocketAddress) ctx.channel().remoteAddress());
+      } catch (IllegalStateException ex) {
+        log.log(Level.FINE, "Did not find exactly one filter chain.");
```

Comment:
`ex.getMessage()` has more specific error than "Did not find exactly one filter chain.". How about using that string here?